stream/src/public_key/connection.rs

```diff
@@ -337,14 +337,19 @@ impl<St: Stream> crate::Receiver for Receiver<St> {
         )
         .await?;
 
-        // Decrypt data
-        let msg = self
-            .cipher
-            .decrypt(&self.nonce.encode(), msg.as_ref())
-            .map_err(|_| Error::DecryptionFailed)?;
+        // Encode the nonce before incrementing to get the proper value
+        let nonce = self.nonce.encode();
+
+        // Increment nonce before decrypting so that even if decryption fails, the nonce is still
+        // incremented. Typically, a user of this receiver would want to terminate the connection
+        // anyway on decryption failure, but we do it here in case they don't.
         self.nonce.inc()?;
```

Comment:
Would it be more correct to wait to error on an invalid nonce until we've read the last item? It seems this approach (even in the original code) may lead to us not reading the last item in the stream.

Maybe I'm missing how this works.

Reply:
I think the fix would be "use the incremented nonce to read the message" rather than the current nonce (where we increment after)?